lib/internal/worker.js

```diff
@@ -527,6 +543,79 @@ function eventLoopUtilization(util1, util2) {
   );
 }
 
+function setConnectionsListener(fn) {
+  if (isMainThread && !mainPortWasSetup) {
+    setupMainPort();
+    mainPortWasSetup = true;
+  }
+
+  if (typeof fn === 'undefined' || fn === null) {
+    connectionsListener = null;
+    return;
+  }
+
+  validateFunction(fn, 'fn');
+  connectionsListener = fn;
+}
+
+async function processConnectionRequest(request) {
+  const status = new Int32Array(request.memory);
+
+  try {
+    const result = await connectionsListener?.(request.sender, request.port, request.data);
+    AtomicsStore(status, WORKER_MESSAGING_RESULT_INDEX, result === true ? 0 : 1);
+  } catch (e) {
+    debug('connections listener rejected', e);
+    AtomicsStore(status, WORKER_MESSAGING_RESULT_INDEX, 2);
+  } finally {
+    AtomicsNotify(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+  }
+}
+
+async function connect(target, data) {
+  if (target === threadId) {
+    throw new ERR_WORKER_SAME_THREAD();
+  }
+
+  // Create a shared array to exchange the status and the result
+  const memory = new SharedArrayBuffer(WORKER_MESSAGING_SHARED_DATA);
+  const status = new Int32Array(memory);
+  const promise = AtomicsWaitAsync(status, WORKER_MESSAGING_STATUS_INDEX, 0).value;
+
+  // Create the channel and send it to the other thread
+  const { port1, port2 } = new MessageChannel();
+  sendToWorker(target, { type: messageTypes.CONNECT, sender: threadId, port: port2, memory, data }, [port2]);
+
+  // Wait for the response
+  await promise;
```

Comment:
Is there a risk that the promise never settles?
e.g. `sendToWorker` succeeded to post the message but the targeted worker ends before actually processing the message.

Reply:
Unfortunately yes, but there is no easy way to detect it. Since the `connectionsListener` is also `async`, we can't assume any timeout. I hope this is an edge case.

Reply:
I think we should for document such limitations/edge cases and in special what user can do to avoid them.
Maybe we could allow a user to pass in a timeout value and use `Promise.race()` here.

Reply:
I just realized we already have the timeout support via `Atomics.waitSync`. I just added the parameter.